cid.go

```diff
@@ -26,15 +26,20 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"io"
 	"strings"
 
 	mbase "github.com/multiformats/go-multibase"
 	mh "github.com/multiformats/go-multihash"
+
+	cbg "github.com/whyrusleeping/cbor-gen"
 )
 
 // UnsupportedVersionString just holds an error message
 const UnsupportedVersionString = "<unsupported cid version>"
 
+const CidMaxLen = 256
```

Comment:
will this affect cids using identity multihashes?

Reply:
Yes, do you make identity CIDs that are longer than 256 bytes? If so, that seems rather obscene... but I could be convinced to raise this to 512 bytes.

Reply:
This is where we use it:
https://github.com/Peergos/Peergos/blob/master/src/peergos/shared/crypto/FragmentedPaddedCipherText.java#L71
Non huge directories and files < 4096 bytes are inlined to save a network round trip.

Reply:
That won't be a root object, but inside another ipld object.

Reply:
Any restriction here that is not also enforced on identity multihashes would also mean we could construct Cids that would look fine, but just fail to serialize at runtime.

Reply:
Actually the value is 4112 bytes for that example. Here is an example of a 4096 byte file being encrypted and wrapped in some other cbor (so it includes a 4112 byte identity multihash inlined) - in hex in the file here:

https://alpha.peergos.net/#{"secretLink":true,"link":"#6MDZhRRPT4ugkJuUfceM6bPnpQKEj5dB2NqLxD1RxFn3oA3CusXayN8RReauEh/6MDZhRRPT4ugkJuUfcRzRbPpFimcBNJx2N9TJDnL4W3ETYhwdsWdvgCkXkwipF/JAMW71tpHDbgcrYrb7SkzeoeD4u9KWLR5PpUvPiBkVuK/5Pf7SupJAsP3FoAbmb5rakBBPZDz8zA9K8siNbbZLKnKijW3pWf"}

Reply:
@ianopolous please don't do that. We set the default max to 32 in go-ipfs for a reason.

> Any restriction here that is not also enforced on identity multihashes would also mean we could construct Cids that would look fine, but just fail to serialize at runtime.

You're right, we should be checking this in the `NewCidV1` function as well (and probably when we decode CIDs).

Reply:
@Stebalien I asked back in March, when I implemented this, and was told inlining 4 KiB identity mulithashes was fine. Now all our users are on it, so breaking that (it is a breaking change in go-ipfs) would break every Peergos user. Current go-ipfs happily accepts much larger values.

Reply:
Where did I say that that? Go-ipfs doesn't explicitly _forbid_ this anywhere, it's just a bad idea. I had _thought_ we left something somewhere saying inlining anything over 100 bytes was a bad idea but I can't find that.

Reply:
It was @whyrusleeping I asked. It makes a bunch of stuff a lot faster for us.